smithy-linters/src/main/java/software/amazon/smithy/linters/CamelCaseValidator.java

```diff
@@ -135,14 +157,66 @@ public List<ValidationEvent> validate(Model model) {
+        Set<Shape> serviceClosure = new HashSet<>();
+        for (ServiceShape serviceShape : model.getServiceShapes()) {
+            Walker walker = new Walker(model);
+            walker.iterateShapes(serviceShape).forEachRemaining(serviceClosure::add);
+        }
+        Stream<MemberShape> memberShapesFromServices = serviceClosure.stream()
+                .filter(Shape::isMemberShape)
+                .map(shape -> (MemberShape) shape);
+
+        events.addAll(validateCamelCasing(memberShapesFromServices.collect(Collectors.toList())));
+
+        // Next get all other member shapes (ex. trait shape members) and validate per namespace grouping
+        Map<String, List<MemberShape>> memberShapesByNamespace = model.toSet(MemberShape.class).stream()
+                .filter(memberShape -> !serviceClosure.contains(memberShape))
+                .collect(Collectors.groupingBy(
+                        memberShape -> memberShape.getContainer().getNamespace()));
+
+        for (List<MemberShape> memberShapeGrouping : memberShapesByNamespace.values()) {
+            events.addAll(validateCamelCasing(memberShapeGrouping));
+        }
 
         return events;
     }
+
+    private List<ValidationEvent> validateCamelCasing(List<MemberShape> memberShapes) {
+        int upperCamelMemberNamesCount = 0;
+        int lowerCamelMemberNamesCount = 0;
+        Set<MemberShape> nonUpperCamelMemberShapes = new HashSet<>();
+        Set<MemberShape> nonLowerCamelMemberShapes = new HashSet<>();
+
+        for (MemberShape memberShape : memberShapes) {
+            if (MemberNameHandling.UPPER.getRegex().matcher(memberShape.getMemberName()).find()) {
+                upperCamelMemberNamesCount++;
+            } else {
+                nonUpperCamelMemberShapes.add(memberShape);
+            }
+            if (MemberNameHandling.LOWER.getRegex().matcher(memberShape.getMemberName()).find()) {
+                lowerCamelMemberNamesCount++;
+            } else {
+                nonLowerCamelMemberShapes.add(memberShape);
+            }
+        }
+
+        // Member shapes are expected to be either upper or lower, depending on the config (and in AUTO mode, the model)
+        Set<MemberShape> violatingMemberShapes = new HashSet<>();
+        if (MemberNameHandling.AUTO.equals(config.getMemberNames())) {
+            violatingMemberShapes = upperCamelMemberNamesCount > lowerCamelMemberNamesCount
+                    ? nonUpperCamelMemberShapes : nonLowerCamelMemberShapes;
+        } else if (MemberNameHandling.UPPER.equals(config.getMemberNames())) {
+            violatingMemberShapes = nonUpperCamelMemberShapes;
+        } else if (MemberNameHandling.LOWER.equals(config.getMemberNames())) {
+            violatingMemberShapes = nonLowerCamelMemberShapes;
+        }
+
+        return violatingMemberShapes.stream()
+                .map(shape -> danger(shape, format(
+                        "Member shape member name, `%s`, is not %s camel case. "
```

Comment:
I think the error message should be more specific. For example, if set to AUTO, then say in this error message what AUTO resolved to whether it was uppercase or lowercase. It would be nice if we also said if this was a service closure or namespace too. For example:

For a service:

```
`UpperCamel` is not lower camel case; members in the FooService must all use lower camel case.
```

And for a namespace:

```
`UpperCamel` is not lower camel case; members in the com.baz namespace must all use lower camel case.
```